src/coreclr/debug/di/rsappdomain.cpp

```diff
@@ -416,7 +416,7 @@ CordbAssembly * CordbAppDomain::CacheAssembly(VMPTR_Assembly vmAssembly)
 {
     INTERNAL_API_ENTRY(GetProcess());
 
-    RSInitHolder<CordbAssembly> pAssembly(new CordbAssembly(this, vmAssembly, VMPTR_DomainAssembly()));
+    RSInitHolder<CordbAssembly> pAssembly(new CordbAssembly(this, vmAssembly, VMPTR_Assembly()));
```

Comment:
I'm not too sure about `vmAssembly->SetHostPtr(pAssembly);`, since later it reads from GetDacPtr/GetRawPtr as well?

Reply:
@elinor-fung, I think the confusion is if we delete `CacheRootAssembly`, then where would we call `vmAssembly->SetHostPtr(pAssembly);` (which basically what `GetAssemblyFromRuntimeAssembly` does)? Going by the number of places `GetDacPtr` and `GetRawPtr` are called suggests that we would need to do more refactoring to straighten it out. Shall we keep this DAC ptr 'self reference' for now?

Reply:
> I think we can remove CacheRootAssembly above and remove the second VMPTR_Assembly parameter and member on CordbAssembly, since it should be the same as the first.

Even without DomainAssembly existing any more, CacheAssembly and CacheRootAssembly still have intentionally different behavior because ICorDebug's domain-neutral assembly concept hasn't been removed yet. CacheRootAssembly creates the normal AppDomain-bound CordbAssembly object that is observed in most of the debugging API. This CordbAssembly previously has one VMPTR_Assembly field and one VMPTR_DomainAssembly field (m_vmDomainAssembly), neither of which is NULL. Contrast this with CacheAssembly() which creates ICorDebug's domain-neutral representation of the same assembly. That CordbAssembly object has a valid VMPTR_Assembly for the first field but it has m_vmDomainAssembly = NULL. The domain-bound and domain-neutral versions of the object have some minor differences in behavior because of m_vmDomainAssembly being NULL or not and we need to preserve it if we are solely claiming this is a refactor.

I think the ideal thing to do is get rid of ICorDebug's domain-neutral assembly concept entirely, but I recommend doing it in a separate PR that we would describe as a behavior change, not a refactor. In particular ICorDebugProcess::GetTypeFromTypeID and GetTypeFromObject should cease producing domain-neutral types and start producing domain affinitized types instead. Once that is done then CacheAssembly should be a dead code path that can be removed.